rasa/core/policies/ensemble.py

```diff
@@ -538,49 +183,56 @@ def _pick_best_policy(
             if confidence > best_confidence:
                 # pick the best policy
                 best_confidence = confidence
-                best_policy_name = policy_name
-
-        best_prediction = predictions.get(best_policy_name)
+                best_index = idx
 
-        if not best_prediction:
+        if best_index < 0:
             raise InvalidConfigException(
-                f"No prediction for policy '{best_policy_name}' found. Please check "
-                f"your model configuration."
+                f"No best prediction found. Please check " f"your model configuration."
             )
 
+        best_prediction = predictions[best_index]
         policy_events += best_prediction.optional_events
 
         return PolicyPrediction(
             best_prediction.probabilities,
-            best_policy_name,
+            best_prediction.policy_name,
             best_prediction.policy_priority,
             policy_events,
             is_end_to_end_prediction=best_prediction.is_end_to_end_prediction,
             is_no_user_prediction=best_prediction.is_no_user_prediction,
             diagnostic_data=best_prediction.diagnostic_data,
             hide_rule_turn=best_prediction.hide_rule_turn,
+            action_metadata=best_prediction.action_metadata,
         )
 
+    @staticmethod
     def _best_policy_prediction(
-        self,
+        predictions: List[PolicyPrediction],
         tracker: DialogueStateTracker,
         domain: Domain,
-        interpreter: NaturalLanguageInterpreter,
     ) -> PolicyPrediction:
         """Finds the best policy prediction.
 
         Args:
-            tracker: the :class:`rasa.core.trackers.DialogueStateTracker`
-            domain: the :class:`rasa.shared.core.domain.Domain`
-            interpreter: Interpreter which may be used by the policies to create
-                additional features.
+            predictions: a list of policy predictions that include "confidence scores"
+              which are non-negative but *do not* necessarily up to 1
+            tracker: dialogue state tracker holding the state of the conversation,
+              which may influence the combination of predictions as well
+            domain: the common domain
 
         Returns:
             The winning policy prediction.
         """
-        # find rejected action before running the policies
-        # because some of them might add events
-        rejected_action_name = None
+        # NOTE: previously this method did find rejected action before running the
+        # policies "because some of them might add events"
+        # - if this comment is obsolete and no prediction modifies trackers during
+        #   predictions, then the following is fine
+        # - if this comment is not obsolete then we have to invent a new
```

Comment:
I don't understand why this would need to be done in a different component?

Reply:
Because this new component works with predictions and hence all policies have run "predict" already. The old code tried to identify the rejected action *before* the policies run predict "because some of them might add events". => We can't do that inside the component that runs after all predicts.

Reply:
Reference: [in this line the "predict" happened - after the rejected event has been extracted](https://github.com/RasaHQ/rasa/blob/main/rasa/core/policies/ensemble.py#L599)

Reply:
Is it possible to remove the events retroactively?

Reply:
If there was a way to distinguish between the events that the policies might add during predict and the ones that where there before, then yep. Can we distinguish that by the timestamp? 

But ideally this is an obsolete comment from a time where we did not have `PolicyPredictions` that could hold events. 

As far as I see:
- TED only uses tracker.latest_action_name and nothing else
- unexpecTEDintent uses past_states and some other getter from the tracker
- memoization modifies trackers - but only "init_copies" of trackers and those copies are never passed on / do not replace the actual tracker
- rule policy only gets information - or modifies an "init_copy" that is *not* passed on
and I expect Johannes to confirm that in the review (which is why I implemented that action rejected stuff here - and left the comments so that he is aware and double checks)

Reply:
( Oh and if that is not the case and all policies could modify the tracker any time they want during predict, then (in addition to that extra component) the predict graph would have to look sequential and pass tracker from one policy to the next in the same order as the for loop in the ensemble before if we wanted to replicate that old behaviour 1:1, I guess 🤔  )

Reply:
To the best of my knowledge, no policy actually modifies the tracker directly. But it would be good to write some tests for the Policy class that checks this. Maybe a customer has a custom policy that does such shenanigans? I don't think we should allow it.

Reply:
That's a good point. We should add tests for that.  -- Agree that we shouldn't allow it. Those side effects would mess with all the nice things we could do with the graph.